src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/PackageSourceMappingUtility.cs

```diff
@@ -12,53 +12,80 @@ namespace NuGet.PackageManagement.UI
             PackageSourceMappingProvider sourceMappingProvider,
-            IReadOnlyList<PackageSourceMappingSourceItem> existingPackageSourceMappingSourceItems)
+            IReadOnlyList<PackageSourceMappingSourceItem> existingPackageSourceMappingSourceItems,
+            out int? countCreatedTopLevelSourceMappings,
+            out int? countCreatedTransitiveSourceMappings)
         {
+            countCreatedTopLevelSourceMappings = null;
+            countCreatedTransitiveSourceMappings = null;
+
             if (userAction?.SelectedSourceName is null || addedPackageIds is null)
             {
                 return;
             }
 
+            countCreatedTopLevelSourceMappings = 0;
+            countCreatedTransitiveSourceMappings = 0;
+
+            string topLevelPackageId = userAction.PackageId;
             Dictionary<string, IReadOnlyList<string>> patternsReadOnly = existingPackageSourceMappingSourceItems
                 .ToDictionary(pair => pair.Key, pair => (IReadOnlyList<string>)(pair.Patterns.Select(p => p.Pattern).ToList()));
 
             PackageSourceMapping packageSourceMapping = new(patternsReadOnly);
 
             // Expand all patterns/globs so we can later check if this package ID was already mapped.
-            List<string> addedPackageIdsWithoutExistingMappings = new();
+            List<string> addedPackageIdsWithoutExistingMappings = new(capacity: addedPackageIds.Count + 1);
+
             foreach (string addedPackageId in addedPackageIds)
             {
                 IReadOnlyList<string> configuredSource = packageSourceMapping.GetConfiguredPackageSources(addedPackageId);
-                if (configuredSource.Count == 0)
+
+                // Top-level package was looked up.
+                if (addedPackageId == topLevelPackageId)
+                {
+                    // The top-level package is not already mapped to the selected source.
+                    if (configuredSource.Count == 0 || !configuredSource.Contains(userAction.SelectedSourceName))
+                    {
+                        countCreatedTopLevelSourceMappings = 1;
+                        addedPackageIdsWithoutExistingMappings.Add(topLevelPackageId);
+                    }
+                }
+                // Transitive package was looked up.
+                else if (configuredSource.Count == 0)
                 {
                     addedPackageIdsWithoutExistingMappings.Add(addedPackageId);
                 }
             }
 
-            // Get all newly added package IDs that were not previously Source Mapped.
-            // Always include the Package ID being installed since it takes precedence over any globbing.
-            string[] packageIdsNeedingNewSourceMappings = addedPackageIdsWithoutExistingMappings
-                .Append(userAction.PackageId)
-                .ToArray();
+            countCreatedTransitiveSourceMappings = addedPackageIdsWithoutExistingMappings.Count - countCreatedTopLevelSourceMappings;
```

Comment:
The problem with this calculation is that, if there's a bug in the logic in this function, telemetry would not help us understand what the bug is. Instead, I suggest incrementing the counter after line 61 (within the else if) and reporting that.